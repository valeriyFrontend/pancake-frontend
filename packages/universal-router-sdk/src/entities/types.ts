import { ActionsPlanner } from '@pancakeswap/infinity-sdk'
import { PermitSingle } from '@pancakeswap/permit2-sdk'
import { BigintIsh, Currency, CurrencyAmount, TradeType } from '@pancakeswap/sdk'
import { BaseRoute, Pool, RouteType, SmartRouterTrade, SwapOptions } from '@pancakeswap/smart-router'
import { Address } from 'viem'

export interface Permit2Signature extends PermitSingle {
  signature: `0x${string}`
}

export type SwapRouterConfig = {
  sender?: Address // address
  deadline?: BigintIsh | undefined
}

export type FlatFeeOptions = {
  amount: BigintIsh
  recipient: Address
}

export type PaymentOptions = {
  /**
   * Whether the payer of the trade is the user. Defaults to true.
   */
  payerIsUser?: boolean
}

export type PancakeSwapOptions = Omit<SwapOptions, 'inputTokenPermit'> & {
  inputTokenPermit?: Permit2Signature
  flatFee?: FlatFeeOptions
} & PaymentOptions

export type SwapRouteMeta = {
  maximumAmountIn: CurrencyAmount<Currency>
  minimumAmountOut: CurrencyAmount<Currency>
  sections: SwapSection[]
}

export type SwapSection = {
  type: RouteType.V2 | RouteType.V3 | RouteType.InfinityBIN | RouteType.InfinityCL | RouteType.STABLE
  poolIn: Currency
  poolOut: Currency
  wrapInput: boolean
  unwrapInput: boolean
  wrapOutput: boolean
  unwrapOutput: boolean
  isFirstSection: boolean
  isLastSection: boolean
  payerIsUser: boolean
  pools: Pool[]
  route: BaseRoute
  nextSection: SwapSection | null
  prevSection: SwapSection | null
  isInfinity: boolean
}

export type SwapTradeContext = {
  trade: Omit<SmartRouterTrade<TradeType>, 'gasEstimate'>
  options: PancakeSwapOptions
  routes: SwapRouteMeta[]
  mustCustody: boolean
  user: Address

  /* null means no changes needs return, wrap=true means needs wrap, wrap=false means needs unwrap */
  returnChanges: {
    wrap: boolean
  } | null

  /**
   * Take over the wrap/unwrap process in last stage
   */
  takeOverWrapSweep: {
    wrap: boolean
  } | null

  /* Should merge wrap ETH before each swap */
  mergedWrapBeforeTrade: {
    wrap: boolean
  } | null

  /*
   * Reuse ActionsPlaner when consecutive infinity sections
   */
  reusedActionPlaner: ActionsPlanner | null
}
