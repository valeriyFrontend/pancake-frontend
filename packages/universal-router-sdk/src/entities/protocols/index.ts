export * from './TradePlanner'
export * from './unwrapWETH'
