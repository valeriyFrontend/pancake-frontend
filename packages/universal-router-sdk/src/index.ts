export { getUniversalRouterAddress } from './constants'
export * from './entities'
export * from './entities/types'
export { PancakeSwapUniversalRouter } from './PancakeSwapUniversalRouter'
