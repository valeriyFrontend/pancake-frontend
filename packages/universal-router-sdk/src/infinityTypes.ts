import { Bytes32, EncodedPoolKey } from '@pancakeswap/infinity-sdk'
import { Address } from 'viem'

export interface PathKey {
  intermediateCurrency: Address
  fee: number
  hooks: Address
  poolManager: Address
  hookData: `0x${string}`
  parameters: `0x${string}`
}

export interface PoolKey {
  currency0: Address
  currency1: Address
  hooks: Address
  poolManager: Address
  fee: number
  parameters: `0x${string}`
}

// export interface CLSwapExactInputParams {
//   address: Address
//   path: PathKey[]
//   amountIn: bigint
//   amountOutMinimum: bigint
// }

// export interface CLSwapExactInputSingleParams {
//   poolKey: PoolKey
//   zeroForOne: boolean
//   amountIn: bigint
//   amountOutMinimum: bigint
//   hookData: `0x${string}`
// }

// export interface BinSwapExactInputSingleParams {
//   poolKey: PoolKey
//   swapForY: boolean
//   amountIn: bigint
//   amountOutMinimum: bigint
//   hookData: `0x${string}`
// }

export interface EncodedSingleSwapParams {
  poolKey: EncodedPoolKey
  zeroForOne: boolean
  hookData: Bytes32
}

export interface EncodedExactInputParams {
  amountIn: bigint
  amountOutMinimum: bigint
}
export interface EncodedExactOutputParams {
  amountOut: bigint
  amountInMaximum: bigint
}

export interface EncodedSingleSwapInParams extends EncodedSingleSwapParams, EncodedExactInputParams {}
export interface EncodedSingleSwapOutParams extends EncodedSingleSwapParams, EncodedExactOutputParams {}

export type EncodedPathKey = {
  intermediateCurrency: Address
  fee: number
  hooks: Address
  poolManager: Address
  hookData: `0x${string}`
  parameters: `0x${string}`
}

export interface EncodedMultiSwapInParams extends EncodedExactInputParams {
  currencyIn: Address
  path: EncodedPathKey[]
}
export interface EncodedMultiSwapOutParams extends EncodedExactOutputParams {
  currencyOut: Address
  path: EncodedPathKey[]
}
