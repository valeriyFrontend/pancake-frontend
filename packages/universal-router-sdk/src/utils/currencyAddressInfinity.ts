import { Currency, ZERO_ADDRESS } from '@pancakeswap/sdk'

export function currencyAddressInfinity(currency: Currency) {
  return currency.isNative ? ZERO_ADDRESS : currency.wrapped.address
}
