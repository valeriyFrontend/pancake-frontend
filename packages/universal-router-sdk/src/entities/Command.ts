import { RoutePlanner } from '../utils/RoutePlanner'
import { PaymentOptions } from './types'

export type TradeConfig = PaymentOptions & {
  allowRevert: boolean
}

export enum RouterTradeType {
  PancakeSwapTrade = 'PancakeSwapTrade',
  // NFTTrade = 'NFTTrade',
  UnwrapWETH = 'UnwrapWETH',
}

// interface for entities that can be encoded as a Universal Router command
export interface Command {
  tradeType: RouterTradeType
  encode(planner: RoutePlanner, config: PaymentOptions): void
}
