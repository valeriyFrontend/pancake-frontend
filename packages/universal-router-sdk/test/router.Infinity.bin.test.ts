import { ChainId } from '@pancakeswap/chains'
import { ACTION_CONSTANTS, ACTIONS, decodePoolKey } from '@pancakeswap/infinity-sdk'
import { CurrencyAmount, ERC20Token, Ether, Percent, TradeType, ZERO_ADDRESS } from '@pancakeswap/sdk'
import { InfinityBinPool, MSG_SENDER, SmartRouter } from '@pancakeswap/smart-router'
import { ADDRESS_ZERO } from '@pancakeswap/v3-sdk'
import { isHex, parseEther, stringify } from 'viem'
import { beforeEach, describe, expect, it } from 'vitest'
import { PancakeSwapUniversalRouter } from '../src'
import { PancakeSwapOptions } from '../src/entities/types'
import { EncodedMultiSwapInParams, EncodedMultiSwapOutParams, EncodedSingleSwapOutParams } from '../src/infinityTypes'
import { CommandType } from '../src/router.types'
import { decodeUniversalCalldata } from '../src/utils/calldataDecode'
import { currencyAddressInfinity } from '../src/utils/currencyAddressInfinity'
import { fixtureAddresses } from './fixtures/address'
import { buildInfinityTrade } from './utils/buildTrade'
import { testInfinitySettleAction, testInfinityTakeAction } from './utils/commandTestUtils'

const TEST_FEE_RECIPIENT_ADDRESS = '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB' as const

const swapOptions = (options: Partial<PancakeSwapOptions>): PancakeSwapOptions => {
  let slippageTolerance = new Percent(5, 100)
  if (options.fee) slippageTolerance = slippageTolerance.add(options.fee.fee)
  return {
    slippageTolerance,
    ...options,
  }
}

const TEST_FEE = 500n

const feeOptions = {
  recipient: TEST_FEE_RECIPIENT_ADDRESS,
  fee: new Percent(TEST_FEE, 10000n),
}

describe('PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test', () => {
  const chainId = ChainId.ETHEREUM
  const liquidity = parseEther('1000')

  let ETHER: Ether
  let USDC: ERC20Token
  let CAKE: ERC20Token
  let ETH_CAKE_BIN_INFI: InfinityBinPool
  let ETH_USDC_BIN_INFI: InfinityBinPool

  expect.addSnapshotSerializer({
    serialize(val) {
      return stringify(decodeUniversalCalldata(val), null, 2)
    },
    test(val) {
      return val && isHex(val)
    },
  })

  beforeEach(async () => {
    ;({ ETHER, CAKE, USDC, ETH_CAKE_BIN_INFI, ETH_USDC_BIN_INFI } = await fixtureAddresses(chainId, liquidity))
  })

  describe('Infinity-BIN', () => {
    it('should encode a single exactInput ETH-CAKE BIN swap zeroForOne', async () => {
      const amountIn = parseEther('0.01')
      const inputAmount = CurrencyAmount.fromRawAmount(ETHER, amountIn)
      const outputAmount = CurrencyAmount.fromRawAmount(CAKE, parseEther('1'))
      const trade = buildInfinityTrade(TradeType.EXACT_INPUT, inputAmount, outputAmount, [ETH_CAKE_BIN_INFI])

      const options = swapOptions({})

      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()
      const maxIn = SmartRouter.maximumAmountIn(trade, options.slippageTolerance)

      expect(BigInt(value)).toEqual(amountIn)

      const decodedCommands = decodeUniversalCalldata(calldata)

      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      expect(decodedCommands[0].args.length).toEqual(0)
      const actions = decodedCommands[0].actions!

      // CL_SWAP_EXACT_IN_SINGLE
      expect(actions[0].action).toEqual(ACTIONS[ACTIONS.BIN_SWAP_EXACT_IN_SINGLE])
      expect(actions[0].args[0].name).toEqual('params')
      const encoded = actions[0].args[0].value as any
      const poolKey = decodePoolKey(encoded.poolKey, 'Bin')
      expect(poolKey.currency0).toEqual(currencyAddressInfinity(inputAmount.currency))
      expect(poolKey.currency1).toEqual(currencyAddressInfinity(outputAmount.currency))

      testInfinitySettleAction(actions[1], ETHER, maxIn.quotient, true)
      testInfinityTakeAction(actions[2], CAKE, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })

    it('should encode a single exactInput CAKE->ETH BIN swap oneForZero', async () => {
      const inputAmount = CurrencyAmount.fromRawAmount(CAKE, 10000n)
      const outputAmount = CurrencyAmount.fromRawAmount(ETHER, 50n)
      const trade = buildInfinityTrade(TradeType.EXACT_INPUT, inputAmount, outputAmount, [ETH_CAKE_BIN_INFI])

      const options = swapOptions({})

      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()

      expect(BigInt(value)).toEqual(0n)

      const decodedCommands = decodeUniversalCalldata(calldata)

      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      expect(decodedCommands[0].args.length).toEqual(0)
      const actions = decodedCommands[0].actions!

      // CL_SWAP_EXACT_IN_SINGLE
      expect(actions[0].action).toEqual(ACTIONS[ACTIONS.BIN_SWAP_EXACT_IN_SINGLE])
      expect(actions[0].args[0].name).toEqual('params')
      const encoded = actions[0].args[0].value as any
      const poolKey = decodePoolKey(encoded.poolKey, 'CL')
      expect(poolKey.currency1).toEqual(currencyAddressInfinity(inputAmount.currency))
      expect(poolKey.currency0).toEqual(currencyAddressInfinity(outputAmount.currency))

      // SETTLE_ALL
      testInfinitySettleAction(actions[1], CAKE, inputAmount.quotient, true)

      // TAKE
      testInfinityTakeAction(actions[2], ETHER, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })

    it('should encode a single exactOutput ETH_CAKE BIN swap zeroForOne', async () => {
      const inputAmount = CurrencyAmount.fromRawAmount(ETHER, 50)
      const outputAmount = CurrencyAmount.fromRawAmount(CAKE, 10000)
      const trade = buildInfinityTrade(TradeType.EXACT_OUTPUT, inputAmount, outputAmount, [ETH_CAKE_BIN_INFI])

      const options = swapOptions({})

      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()

      // const minOut = SmartRouter.minimumAmountOut(trade, options.slippageTolerance)
      const maximumAmountIn = SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient

      expect(BigInt(value)).toEqual(maximumAmountIn)
      // expect(calldata).toMatchSnapshot()

      const decodedCommands = decodeUniversalCalldata(calldata)

      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      expect(decodedCommands[0].args.length).toEqual(0)
      const actions = decodedCommands[0].actions!

      // CL_SWAP_EXACT_IN_SINGLE
      expect(actions[0].action).toEqual(ACTIONS[ACTIONS.BIN_SWAP_EXACT_OUT_SINGLE])
      expect(actions[0].args[0].name).toEqual('params')
      const swapParams = actions[0].args[0].value as EncodedSingleSwapOutParams

      const poolKey = decodePoolKey(swapParams.poolKey, 'CL')
      expect(poolKey.currency0).toEqual(currencyAddressInfinity(inputAmount.currency))
      expect(poolKey.currency1).toEqual(currencyAddressInfinity(outputAmount.currency))
      expect(swapParams.amountOut).toEqual(outputAmount.quotient)
      expect(swapParams.amountInMaximum).toEqual(maximumAmountIn)

      testInfinitySettleAction(actions[1], ETHER, maximumAmountIn, true)
      testInfinityTakeAction(actions[2], CAKE, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
      testInfinityTakeAction(actions[3], ETHER, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })

    it('should encode a single exactOutput CAKE-ETH BIN swap OneForZero', async () => {
      const inputAmount = CurrencyAmount.fromRawAmount(CAKE, 10000)
      const outputAmount = CurrencyAmount.fromRawAmount(ETHER, 30)
      const trade = buildInfinityTrade(TradeType.EXACT_OUTPUT, inputAmount, outputAmount, [ETH_CAKE_BIN_INFI])

      const options = swapOptions({})
      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()
      const maximumAmountIn = SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient

      expect(BigInt(value)).toEqual(0n)

      const decodedCommands = decodeUniversalCalldata(calldata)
      expect(decodedCommands.length).toEqual(1)
      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      const actions = decodedCommands[0].actions!
      expect(actions.length).toEqual(4)

      // ACTION #0
      const swapParams = actions[0].args[0].value as EncodedSingleSwapOutParams
      expect(swapParams.zeroForOne).toEqual(false)
      expect(swapParams.poolKey.currency0).toEqual(ADDRESS_ZERO)
      expect(swapParams.poolKey.currency1).toEqual(inputAmount.currency.address)
      expect(swapParams.amountOut).toEqual(outputAmount.quotient)
      expect(swapParams.amountInMaximum).toEqual(maximumAmountIn)

      // ACTION #1
      testInfinitySettleAction(actions[1], CAKE, maximumAmountIn, true)
      testInfinityTakeAction(actions[2], ETHER, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
      testInfinityTakeAction(actions[3], CAKE, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })

    it('should encode a multi-hop exactInput in Infinity BIN: CAKE->ETH->USDC', async () => {
      const inputAmount = CurrencyAmount.fromRawAmount(CAKE, 10000)
      const outputAmount = CurrencyAmount.fromRawAmount(USDC, 1000)

      const trade = buildInfinityTrade(TradeType.EXACT_INPUT, inputAmount, outputAmount, [
        ETH_CAKE_BIN_INFI,
        ETH_USDC_BIN_INFI,
      ])

      const options = swapOptions({})
      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()
      const minOutAmount = SmartRouter.minimumAmountOut(trade, options.slippageTolerance).quotient

      expect(BigInt(value)).toEqual(0n)
      const decodedCommands = decodeUniversalCalldata(calldata)

      expect(decodedCommands.length).toEqual(1)
      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      const actions = decodedCommands[0].actions!
      expect(actions.length).toEqual(3)

      // ACTION #0
      const swapParams = actions[0].args[0].value as EncodedMultiSwapInParams
      expect(swapParams.currencyIn).toEqual(CAKE.address)
      expect(swapParams.amountIn).toEqual(10000n)
      expect(swapParams.amountOutMinimum).toEqual(minOutAmount)

      //  PATH[0]
      const path0 = swapParams.path[0]
      expect(path0.intermediateCurrency).toEqual(ZERO_ADDRESS)

      // PATH[1]
      const path1 = swapParams.path[1]
      expect(path1.intermediateCurrency).toEqual(USDC.address)

      testInfinitySettleAction(actions[1], CAKE, 10000n, true)
      testInfinityTakeAction(actions[2], USDC, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })

    it('should encode a multi-hop exactOutput in Infinity BIN: USDC->ETH->CAKE', async () => {
      const inputAmount = CurrencyAmount.fromRawAmount(USDC, 1000)
      const outputAmount = CurrencyAmount.fromRawAmount(CAKE, 10000)

      const trade = buildInfinityTrade(TradeType.EXACT_OUTPUT, inputAmount, outputAmount, [
        ETH_USDC_BIN_INFI,
        ETH_CAKE_BIN_INFI,
      ])

      const options = swapOptions({})
      const { calldata, value } = PancakeSwapUniversalRouter.swapERC20CallParameters(trade, options)
      expect(calldata).toMatchSnapshot()
      const maxInAmount = SmartRouter.maximumAmountIn(trade, options.slippageTolerance).quotient

      expect(BigInt(value)).toEqual(0n)
      const decodedCommands = decodeUniversalCalldata(calldata)

      expect(decodedCommands.length).toEqual(1)
      expect(decodedCommands[0].command).toEqual(CommandType[CommandType.INFI_SWAP])
      const actions = decodedCommands[0].actions!

      // ACTION #0
      const swapParams = actions[0].args[0].value as EncodedMultiSwapOutParams
      expect(swapParams.amountInMaximum).toEqual(maxInAmount)
      expect(swapParams.amountOut).toEqual(10000n)

      //  PATH[0]
      const path0 = swapParams.path[0]
      expect(path0.intermediateCurrency).toEqual(USDC.address)

      // PATH[1]
      const path1 = swapParams.path[1]
      expect(path1.intermediateCurrency).toEqual(ZERO_ADDRESS)

      testInfinitySettleAction(actions[1], USDC, maxInAmount, true)
      testInfinityTakeAction(actions[2], CAKE, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
      testInfinityTakeAction(actions[3], USDC, MSG_SENDER, ACTION_CONSTANTS.OPEN_DELTA)
    })
  })
})
