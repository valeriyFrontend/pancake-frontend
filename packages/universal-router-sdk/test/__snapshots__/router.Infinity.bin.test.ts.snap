// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a multi-hop exactInput in Infinity BIN: CAKE->ETH->USDC 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_IN",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "currencyIn": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
              "path": [
                {
                  "intermediateCurrency": "0x0000000000000000000000000000000000000000",
                  "fee": 500,
                  "hooks": "0x0000000000000000000000000000000000000000",
                  "poolManager": "0x0000000000000000000000000000000000000000",
                  "hookData": "0x",
                  "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
                },
                {
                  "intermediateCurrency": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                  "fee": 500,
                  "hooks": "0x0000000000000000000000000000000000000000",
                  "poolManager": "0x0000000000000000000000000000000000000000",
                  "hookData": "0x",
                  "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
                }
              ],
              "amountIn": "10000",
              "amountOutMinimum": "952"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "10000"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a multi-hop exactOutput in Infinity BIN: USDC->ETH->CAKE 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_OUT",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "currencyOut": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
              "path": [
                {
                  "intermediateCurrency": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                  "fee": 500,
                  "hooks": "0x0000000000000000000000000000000000000000",
                  "poolManager": "0x0000000000000000000000000000000000000000",
                  "hookData": "0x",
                  "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
                },
                {
                  "intermediateCurrency": "0x0000000000000000000000000000000000000000",
                  "fee": 500,
                  "hooks": "0x0000000000000000000000000000000000000000",
                  "poolManager": "0x0000000000000000000000000000000000000000",
                  "hookData": "0x",
                  "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
                }
              ],
              "amountOut": "10000",
              "amountInMaximum": "1050"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "1050"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a single exactInput CAKE->ETH BIN swap oneForZero 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_IN_SINGLE",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "poolKey": {
                "currency0": "0x0000000000000000000000000000000000000000",
                "currency1": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
                "hooks": "0x0000000000000000000000000000000000000000",
                "poolManager": "0x0000000000000000000000000000000000000000",
                "fee": 500,
                "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
              },
              "zeroForOne": false,
              "amountIn": "10000",
              "amountOutMinimum": "47",
              "hookData": "0x0000000000000000000000000000000000000000"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "10000"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x0000000000000000000000000000000000000000"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a single exactInput ETH-CAKE BIN swap zeroForOne 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_IN_SINGLE",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "poolKey": {
                "currency0": "0x0000000000000000000000000000000000000000",
                "currency1": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
                "hooks": "0x0000000000000000000000000000000000000000",
                "poolManager": "0x0000000000000000000000000000000000000000",
                "fee": 500,
                "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
              },
              "zeroForOne": true,
              "amountIn": "10000000000000000",
              "amountOutMinimum": "952380952380952380",
              "hookData": "0x0000000000000000000000000000000000000000"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x0000000000000000000000000000000000000000"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "10000000000000000"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a single exactOutput CAKE-ETH BIN swap OneForZero 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_OUT_SINGLE",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "poolKey": {
                "currency0": "0x0000000000000000000000000000000000000000",
                "currency1": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
                "hooks": "0x0000000000000000000000000000000000000000",
                "poolManager": "0x0000000000000000000000000000000000000000",
                "fee": 500,
                "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
              },
              "zeroForOne": false,
              "amountOut": "30",
              "amountInMaximum": "10500",
              "hookData": "0x0000000000000000000000000000000000000000"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "10500"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x0000000000000000000000000000000000000000"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;

exports[`PancakeSwap Universal Router Infinity-Bin Pool Command Generation Test > Infinity-BIN > should encode a single exactOutput ETH_CAKE BIN swap zeroForOne 1`] = `
[
  {
    "command": "INFI_SWAP",
    "args": [],
    "actions": [
      {
        "action": "BIN_SWAP_EXACT_OUT_SINGLE",
        "args": [
          {
            "type": "tuple",
            "name": "params",
            "value": {
              "poolKey": {
                "currency0": "0x0000000000000000000000000000000000000000",
                "currency1": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898",
                "hooks": "0x0000000000000000000000000000000000000000",
                "poolManager": "0x0000000000000000000000000000000000000000",
                "fee": 500,
                "parameters": "0x0000000000000000000000000000000000000000000000000000000000050000"
              },
              "zeroForOne": true,
              "amountOut": "10000",
              "amountInMaximum": "52",
              "hookData": "0x0000000000000000000000000000000000000000"
            }
          }
        ]
      },
      {
        "action": "SETTLE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x0000000000000000000000000000000000000000"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "52"
          },
          {
            "type": "bool",
            "name": "payerIsUser",
            "value": true
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x152649eA73beAb28c5b49B26eb48f7EAD6d4c898"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      },
      {
        "action": "TAKE",
        "args": [
          {
            "type": "address",
            "name": "currency",
            "value": "0x0000000000000000000000000000000000000000"
          },
          {
            "type": "address",
            "name": "recipient",
            "value": "0x0000000000000000000000000000000000000001"
          },
          {
            "type": "uint256",
            "name": "amount",
            "value": "0"
          }
        ]
      }
    ]
  }
]
`;
